src/esp/gfx/SensorInfoVisualizer.cpp

```diff
@@ -0,0 +1,90 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+#include "SensorInfoVisualizer.h"
+
+#include <Corrade/Utility/Assert.h>
+#include <Magnum/GL/DefaultFramebuffer.h>
+#include <Magnum/GL/RenderbufferFormat.h>
+#include <Magnum/Magnum.h>
+#include <Magnum/Math/Vector4.h>
+#include <Magnum/Shaders/Generic.h>
+
+#include "DepthVisualizerShader.h"
+
+namespace esp {
+namespace gfx {
+
+namespace Mn = Magnum;
+namespace Cr = Corrade;
+
+SensorInfoVisualizer::SensorInfoVisualizer() {
+  // prepare a big triangle mesh to cover the screen
+  mesh_ = Mn::GL::Mesh{};
+  mesh_.setCount(3);
+}
+void SensorInfoVisualizer::resetFramebufferRenderbuffer(
+    Mn::Vector2i framebufferSize) {
+  if (frameBuffer_.viewport() != Mn::Range2Di{{}, {framebufferSize}}) {
+    frameBuffer_ = Mn::GL::Framebuffer{{{}, framebufferSize}};
+    colorBuffer_.setStorage(Mn::GL::RenderbufferFormat::RGBA8, framebufferSize);
+    frameBuffer_.attachRenderbuffer(Mn::GL::Framebuffer::ColorAttachment{0},
+                                    colorBuffer_);
+    // depth buffer is 24-bit integer pixel
+    depthBuffer_.setStorage(Mn::GL::RenderbufferFormat::DepthComponent24,
+                            framebufferSize);
+    frameBuffer_.attachRenderbuffer(
+        Mn::GL::Framebuffer::BufferAttachment::Depth, depthBuffer_);
+  }
+}
+
+void SensorInfoVisualizer::prepareToDraw(Magnum::Vector2i framebufferSize) {
+  resetFramebufferRenderbuffer(framebufferSize);
+
+  frameBuffer_.bind();
+  frameBuffer_.mapForDraw({
+      {Mn::Shaders::Generic3D::ColorOutput,
+       Mn::GL::Framebuffer::ColorAttachment{0}},
+  });
+  frameBuffer_.clearDepth(1.0f).clearColor(0,                // color attachment
+                                           Mn::Vector4ui{0}  // clear color
+  );
+
+  CORRADE_INTERNAL_ASSERT(
+      frameBuffer_.checkStatus(Mn::GL::FramebufferTarget::Draw) ==
+      Mn::GL::Framebuffer::Status::Complete);
+}
+
+void SensorInfoVisualizer::blitRgbaToDefault() {
+  frameBuffer_.mapForRead(Mn::GL::Framebuffer::ColorAttachment{0});
+  CORRADE_ASSERT(
+      frameBuffer_.viewport() == Mn::GL::defaultFramebuffer.viewport(),
+      "SensorInfoVisualizer::blitRgbaToDefault(): a mismatch on viewport"
+      "between current framebuffer and the default framebuffer.", );
+
+  Mn::GL::AbstractFramebuffer::blit(
+      frameBuffer_, Mn::GL::defaultFramebuffer, frameBuffer_.viewport(),
+      Mn::GL::defaultFramebuffer.viewport(), Mn::GL::FramebufferBlit::Color,
+      Mn::GL::FramebufferBlitFilter::Nearest);
+}
```

Comment:
Actually, is the extra framebuffer needed *at all*? Since the only use of this visualizer is in the viewer, where you *do want* to have it in the default framebuffer anyway.

I'd personally just have `draw()` take the target framebuffer it renders to and that's it. I don't see a reason why there would need to be yet another render target for visualization operations.

Reply:
See my comments below.